lib/Core/Search/Common/PageTextExtractor.php

```diff
@@ -0,0 +1,269 @@
+    }
+
+    public function setLogger(LoggerInterface $logger): void
+    {
+        $this->logger = $logger;
+    }
+
+    /**
+     * @param int $contentId
+     * @param string $languageCode
+     *
+     * @return array<string, array<int, string>|string>
+     */
+    public function extractPageText(int $contentId, string $languageCode): array
+    {
+        if (isset($this->cache[$contentId][$languageCode])) {
+            return $this->cache[$contentId][$languageCode];
+        }
+
+        if (count($this->cache) > 10) {
+            $this->cache = [];
+        }
+
+        try {
+            $html = $this->fetchPageSource($contentId, $languageCode);
+        } catch (IndexPageUnavailableException|RuntimeException $e) {
+            $this->logger->error($e->getMessage());
+
+            return [];
+        }
+
+        $textArray = $this->extractTextArray($html);
+
+        $this->cache[$contentId][$languageCode] = $textArray;
+
+        return $textArray;
+    }
+
+    /**
+     * @param string $languageCode
+     * @param int $contentId
+     *
+     * @throws \Ibexa\Contracts\Core\Repository\Exceptions\NotFoundException
+     * @throws \Ibexa\Contracts\Core\Repository\Exceptions\UnauthorizedException
+     *
+     * @return string
+     */
+    private function generateUrl(string $languageCode, int $contentId): string
+    {
+        $contentInfo = $this->contentHandler->loadContentInfo($contentId);
+        $siteAccess = $this->resolveSiteAccess($contentInfo, $languageCode);
+
+        $relativePath = $this->router->generate(
+            'ibexa.url.alias',
+            [
+                'locationId' => (int) $contentInfo->mainLocationId,
+                'siteaccess' => $siteAccess,
+            ],
+            UrlGeneratorInterface::RELATIVE_PATH,
+        );
+
+        return $this->pageIndexingHost . $relativePath;
+    }
+
+    private function resolveSiteAccess(ContentInfo $contentInfo, string $languageCode): string
+    {
+        try {
+            $location = $this->locationHandler->load($contentInfo->mainLocationId);
+        } catch (NotFoundException) {
+            throw new RuntimeException(
+                sprintf(
+                    'Content #%d does not have a location',
+                    $contentInfo->id,
+                ),
+            );
+        }
+
+        $pathArray = explode('/', $location->pathString);
+
+        foreach ($this->siteRoots as $site => $siteRoot) {
+            if (!in_array((string) $siteRoot, $pathArray, true)) {
+                continue;
+            }
+
+            if (!isset($this->languageAccessibility[$site][$languageCode])) {
+                throw new RuntimeException("Language not supported for matched siteaccess group '{$site}'");
+            }
+
+            return $this->languageAccessibility[$site][$languageCode];
+        }
+
+        throw new RuntimeException("Failed to match content ID '{$contentInfo->id}' to a siteaccess");
```

Comment:
Use `sprintf` instead of variable interpolation.

Reply:
resolved in https://github.com/netgen/ibexa-search-extra/commit/e497ddad2266d2d86a743bce20942829af26489f